README.rst
============================
repoze.who.plugins.memcached
============================

This is a repoze.who IAuthenticator/IMetadataProvider plugin that can
cache the results of *other* plugins using memcached.  It's useful for
reducing load on e.g. a backend LDAP auth system.

To use it, give it the name of an authenticator and/or metadata provider
whose results it should wrap::

    [plugin:ldap]
    use = my.ldap.authenticator

    [plugin:cached_ldap]
    use = repoze.who.plugins.memcached
    authenticator_name = ldap

    [authenticators]
    plugins = cached_ldap ldap;unused

(The "ldap;unused" bit ensures that the wrapped ldap plugin still gets
loaded, but is not used for matching any requests. Yeah, it's yuck.)

To prevent a compromise of the cache from revealing auth credentials, this
plugin calculates a HMAC hash of the items in the incoming identity and
uses that as the cache key.  This makes it possible to check the cache for
a match to an incoming identity, while preventing the cache keys from being 
reversed back into a valid identity.

Items added to the identity by the wrapped plugin will be stored in the
cached value and will *not* be encryped or obfuscated in any way.

The following configuration options are available:

    * memcached_urls:  A list of URLs for the underlying memcached store.

    * authenticator_name:  The name of an IAuthenticator plugin to wrap.

    * mdprovider_name:  The name of an IMetadataProvider plugin to wrap.

    * key_items:  A list of names from the identity dict that should be
                  hashed to produce the cache key.  These items should
                  uniquely and validly identity a user.  By default it
                  will use all keys in the identity in sorted order.

    * value_items:  A list of names from the identity dict that should be
                    stored in the cache.  These would typically be items
                    of metadata such as the user's email address.  By 
                    default this will include all items that the wrapped
                    plugin adds to the identity.

    * secret:  A string used when calculating the HMAC the cache keys.
               All servers accessing a shared cache should use the same
               secret so they produce the same set of cache keys.

    * ttl:  The time for which cache entries should persist, in seconds.


CHANGES.txt
0.1.1 - 2012-01-30
==================

 * Update license to MPL 2.0


0.1.0 - 2011-12-01
==================

 * Initial release.


repoze/who/plugins/memcached/tests.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is repoze.who.plugins.memcached
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2011
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Ryan Kelly (rkelly@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****

import unittest

import os
import time
import wsgiref.util

from zope.interface import implements
from zope.interface.verify import verifyClass
from repoze.who.config import WhoConfig
from repoze.who.api import APIFactory
from repoze.who.interfaces import (IAuthenticator,
                                   IMetadataProvider,
                                   IIdentifier)

from repoze.who.plugins.memcached import MemcachedPlugin, make_plugin, pylibmc


WHO_CONFIG = """
[plugin:dummy_id]
use = repoze.who.plugins.memcached.tests:DummyIdentifier

[plugin:dummy_auth]
use = repoze.who.plugins.memcached.tests:DummyAuthenticator

[plugin:dummy_md]
use = repoze.who.plugins.memcached.tests:DummyMDProvider

[plugin:cached_auth]
use = repoze.who.plugins.memcached:make_plugin
memcached_urls = localhost:1234
authenticator_name = dummy_auth
mdprovider_name = dummy_md

[identifiers]
plugins = dummy_id

[authenticators]
plugins = cached_auth dummy_auth;unused

[mdproviders]
plugins = cached_auth dummy_md;unused

[general]
challenge_decider = repoze.who.classifiers:default_challenge_decider
request_classifier = repoze.who.classifiers:default_request_classifier
"""


class FakePyLibMCClient(object):
    """Fake pylibmc Client object, for testing purposes."""

    cache = {}

    def __init__(self, urls):
        pass

    def get(self, key):
        try:
            return self.cache[key]
        except KeyError:
            raise pylibmc.ReadError

    def set(self, key, value, time=None):
        self.cache[key] = value

    @classmethod
    def clear_cache(cls):
        cls.cache.clear()


def make_environ(extra=None, **kwds):
    """Construct a wsgi environ dict, with default values."""
    environ = {}
    if extra is not None:
        environ.update(extra)
    environ["wsgi.version"] = (1, 0)
    environ["wsgi.url_scheme"] = "http"
    environ["SERVER_NAME"] = "localhost"
    environ["SERVER_PORT"] = "80"
    environ["REQUEST_METHOD"] = "GET"
    environ["SCRIPT_NAME"] = ""
    environ["PATH_INFO"] = "/"
    environ.update(kwds)
    return environ


class DummyIdentifier(object):
    """IIdentifier that extracts all X-ID-* headers."""

    implements(IIdentifier)

    def identify(self, environ):
        identity = {}
        for header in environ:
            if header.startswith("HTTP_X_ID_"):
                key = header[10:].lower()
                identity[key] = environ[header]
        return identity

    def remember(self, environ, identity):
        raise RuntimeError("tests shouldn't call this")  # pragma: nocover

    def forget(self, environ, identity):
        raise RuntimeError("tests shouldn't call this")  # pragma: nocover


class DummyAuthenticator(object):
    """IAuthenticator with an on/off switch that can add arbitrary keys."""

    implements(IAuthenticator)

    def __init__(self):
        self.active = True
        self.extras = {}

    def authenticate(self, environ, identity):
        if not self.active:
            return None
        identity.update(self.extras)
        return identity.get("username")


class DummyMDProvider(object):
    """IMetadataprovider with an on/off switch that can add arbitrary keys."""

    implements(IMetadataProvider)

    def __init__(self):
        self.active = True
        self.extras = {}

    def add_metadata(self, environ, identity):
        if self.active:
            identity.update(self.extras)


class TestMemcachedPlugin(unittest.TestCase):

    def setUp(self):
        self._orig_Client = pylibmc.Client
        pylibmc.Client = FakePyLibMCClient
        pylibmc.Client.clear_cache()

    def tearDown(self):
        pylibmc.Client.clear_cache()
        pylibmc.Client = self._orig_Client
        del self._orig_Client

    def _make_api_factory(self):
        parser = WhoConfig("")
        parser.parse(WHO_CONFIG)
        return APIFactory(parser.identifiers,
                          parser.authenticators,
                          parser.challengers,
                          parser.mdproviders,
                          parser.request_classifier,
                          parser.challenge_decider)

    def test_implements(self):
        verifyClass(IAuthenticator, MemcachedPlugin)
        verifyClass(IMetadataProvider, MemcachedPlugin)

    def test_make_plugin(self):
        # Test that arguments get parsed and set appropriately.
        plugin = make_plugin(memcached_urls="127.0.0.2:3 127.0.0.1:4",
                             authenticator_name="myauth",
                             mdprovider_name="mymd",
                             key_items="one two three",
                             value_items="five",
                             secret="Ted Koppel is a robot",
                             ttl="42")
        self.assertEquals(plugin.memcached_urls,
                          ["127.0.0.2:3", "127.0.0.1:4"])
        self.assertEquals(plugin.authenticator_name, "myauth")
        self.assertEquals(plugin.mdprovider_name, "mymd")
        self.assertEquals(plugin.key_items, ["one", "two", "three"])
        self.assertEquals(plugin.value_items, ["five"])
        self.assertEquals(plugin.secret, "Ted Koppel is a robot")
        self.assertEquals(plugin.ttl, 42)
        # Test that you must specify memcached urls and a plugin.
        self.assertRaises(TypeError, make_plugin)
        self.assertRaises(ValueError, make_plugin, "")
        self.assertRaises(ValueError, make_plugin, "localhost", secret="HA")
        # Test that appropriate defaults get set.
        plugin = make_plugin("127.0.0.2:3", "auther")
        self.assertEquals(plugin.memcached_urls, ["127.0.0.2:3"])
        self.assertEquals(plugin.authenticator_name, "auther")
        self.assertEquals(plugin.mdprovider_name, None)
        self.assertEquals(plugin.key_items, None)
        self.assertEquals(plugin.value_items, None)
        self.failUnless(isinstance(plugin.secret, basestring))
        self.assertEquals(plugin.ttl, 60)
        # Test setting just the authenticator
        plugin = make_plugin(memcached_urls="127.0.0.2:3",
                             authenticator_name="auther")
        self.assertEquals(plugin.authenticator_name, "auther")
        self.assertEquals(plugin.mdprovider_name, None)
        # Test setting just the mdprovider
        plugin = make_plugin(memcached_urls="127.0.0.2:3",
                             mdprovider_name="mdprov")
        self.assertEquals(plugin.authenticator_name, None)
        self.assertEquals(plugin.mdprovider_name, "mdprov")

    def test_basic_caching_of_login(self):
        api_factory = self._make_api_factory()
        # If we disable the backing plugin then we can't authenticate.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate(), None)
        # Re-enable it, we can login and have our credentials cached.
        api_factory.authenticators[1][1].active = True
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # Now if we disable it, we can still login using the cached values.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # But some other random user can't log in.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test2"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate(), None)

    def test_basic_caching_of_metadata(self):
        api_factory = self._make_api_factory()
        api_factory.mdproviders[1][1].extras = {"meta": "data"}
        # Prime the cache for the given user.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("meta"), "data")
        # Now switch off the mdprovider, we can still use the cache.
        api_factory.mdproviders[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        self.assertEquals(api.authenticate().get("meta"), "data")
        # Uncached users don't get the extra metadata.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test2"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test2")
        self.assertEquals(api.authenticate().get("meta"), None)

    def test_authenticator_can_be_disabled(self):
        api_factory = self._make_api_factory()
        api_factory.authenticators[0][1].authenticator_name = None
        # Log in will fail due to no wrapped authenticator.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate(), None)

    def test_mdprovider_can_be_disabled(self):
        api_factory = self._make_api_factory()
        api_factory.authenticators[0][1].mdprovider_name = None
        api_factory.authenticators[1][1].extras = {"auth": "enticator"}
        api_factory.mdproviders[1][1].extras = {"meta": "data"}
        # Logins will still work, but the mdprovider is not called.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        self.assertEquals(api.authenticate().get("auth"), "enticator")
        self.assertEquals(api.authenticate().get("meta"), None)

    def test_graceful_handling_of_non_repoze_environ(self):
        api_factory = self._make_api_factory()
        api_factory.mdproviders[1][1].extras = {"meta": "data"}
        # Log in will fail due to no repoze.who api in the environ.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        del environ["repoze.who.api"]
        self.assertEquals(api.authenticate(), None)
        # The same applies inside add_metadata.
        identity = {}
        api_factory.mdproviders[0][1].add_metadata(environ, identity)
        self.assertEquals(identity, {})

    def test_graceful_handling_of_cache_write_errors(self):
        api_factory = self._make_api_factory()
        def set_with_error(*args, **kwds):
            raise pylibmc.WriteError
        api_factory.authenticators[0][1]._client.set = set_with_error
        # Log in will succeed despite failed cache write.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # But it won't be cached.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate(), None)

    def test_graceful_handling_of_bad_cache_values(self):
        api_factory = self._make_api_factory()
        # Log in once to prime the cache.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # Now corrupt the cached value.
        cache = api_factory.authenticators[0][1]._client.cache
        for key in cache:
            cache[key] = "I AM NOT VALID JSON"
        # Login attempts with backend disabled will gracefully fail.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate(), None)
        # Login attempts with backend enabled will call through to it.
        api_factory.authenticators[1][1].active = True
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # That will have repaired to cache, so we can use it to login now.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")

    def test_caching_of_extra_values_from_authenticator(self):
        api_factory = self._make_api_factory()
        api_factory.authenticators[1][1].extras = {"one": 1, "two": 2}
        # Log in to let the authenticator add its extra data.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        identity = api.authenticate()
        self.assertEquals(identity.get("one"), 1)
        self.assertEquals(identity.get("two"), 2)
        # Those new values should be restored from the cache.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        identity = api.authenticate()
        self.assertEquals(identity.get("one"), 1)
        self.assertEquals(identity.get("two"), 2)

    def test_caching_of_specific_values_from_authenticator(self):
        api_factory = self._make_api_factory()
        api_factory.authenticators[0][1].value_items = ("one", "three")
        api_factory.authenticators[1][1].extras = {"one": 1, "two": 2}
        # Log in to let the authenticator add its extra data.
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        identity = api.authenticate()
        self.assertEquals(identity.get("one"), 1)
        self.assertEquals(identity.get("two"), 2)
        # Only the specified value should be restored from the cache.
        api_factory.authenticators[1][1].active = False
        environ = make_environ({"HTTP_X_ID_USERNAME": "test1"})
        api = api_factory(environ)
        identity = api.authenticate()
        self.assertEquals(identity.get("one"), 1)
        self.assertEquals(identity.get("two"), None)

    def test_that_cache_key_includes_all_items(self):
        api_factory = self._make_api_factory()
        # Log in with a particular password.
        environ = make_environ(HTTP_X_ID_USERNAME="test1",
                               HTTP_X_ID_PASSWORD="password1")
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # That password will get cached
        api_factory.authenticators[1][1].active = False
        environ = make_environ(HTTP_X_ID_USERNAME="test1",
                               HTTP_X_ID_PASSWORD="password1")
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # But other passwords won't match it, so they fail.
        environ = make_environ(HTTP_X_ID_USERNAME="test1",
                               HTTP_X_ID_PASSWORD="password2")
        api = api_factory(environ)
        self.assertEquals(api.authenticate(), None)

    def test_that_cache_key_excludes_unspecified_items(self):
        api_factory = self._make_api_factory()
        api_factory.authenticators[0][1].key_items = ("username",)
        # Log in with a particular password.
        environ = make_environ(HTTP_X_ID_USERNAME="test1",
                               HTTP_X_ID_PASSWORD="password1")
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # That will get cached using just the username.
        api_factory.authenticators[1][1].active = False
        environ = make_environ(HTTP_X_ID_USERNAME="test1",
                               HTTP_X_ID_PASSWORD="password1")
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")
        # So any old password will hit the same cache value.
        environ = make_environ(HTTP_X_ID_USERNAME="test1",
                               HTTP_X_ID_PASSWORD="password2")
        api = api_factory(environ)
        self.assertEquals(api.authenticate().get("username"), "test1")


repoze/who/plugins/memcached/__init__.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is repoze.who.plugins.memcached
#
# The Initial Developer of the Original Code is the Mozilla Foundation.
# Portions created by the Initial Developer are Copyright (C) 2010
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#   Ryan Kelly (rkelly@mozilla.com)
#   Tarek Ziade (tarek@mozilla.com)
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
"""

repoze.who plugin to cache the results of other plugins.

"""

import os
import json
import hmac
import hashlib

import pylibmc

from zope.interface import implements
from repoze.who.api import get_api
from repoze.who.interfaces import IAuthenticator, IMetadataProvider


class MemcachedPlugin(object):
    """Cache IAuthenticator/IMetadataProvider plugin results using memcached.

    This is a repoze.who IAuthenticator/IMetadataProvider plugin that can
    cache the results of *other* plugins using memcached.  It's useful for
    reducing load on e.g. a backend LDAP auth system.

    To use it, give it the name of an authenticator and/or metadata provider
    whose results it should wrap::

        [plugin:ldap]
        use = my.ldap.authenticator

        [plugin:cached_ldap]
        use = repoze.who.plugins.memcached
        authenticator_name = ldap

        [authenticators]
        plugins = cached_ldap ldap;unused

    (The "ldap;unused" bit ensures that the wrapped ldap plugin still gets
    loaded, but is not used for matching any requests. Yeah, it's yuck.)

    To prevent a compromise of the cache from revealing auth credentials, this
    plugin calculates a HMAC hash of the items in the incoming identity and
    uses that as the cache key.  This makes it possible to check the cache for
    a match to an incoming identity, while preventing the cache keys from being 
    reversed back into a valid identity.

    Items added to the identity by the wrapped plugin will be stored in the
    cached value and will *not* be encryped or obfuscated in any way.

    The following configuration options are available:

        * memcached_urls:  A list of URLs for the underlying memcached store.

        * authenticator_name:  The name of an IAuthenticator plugin to wrap.

        * mdprovider_name:  The name of an IMetadataProvider plugin to wrap.
 
        * key_items:  A list of names from the identity dict that should be
                      hashed to produce the cache key.  These items should
                      uniquely and validly identity a user.  By default it
                      will use all keys in the identity in sorted order.

        * value_items:  A list of names from the identity dict that should be
                        stored in the cache.  These would typically be items
                        of metadata such as the user's email address.  By 
                        default this will include all items that the wrapped
                        plugin adds to the identity.

        * secret:  A string used when calculating the HMAC the cache keys.
                   All servers accessing a shared cache should use the same
                   secret so they produce the same set of cache keys.

        * ttl:  The time for which cache entries should persist, in seconds.

    """

    implements(IAuthenticator, IMetadataProvider)

    def __init__(self, memcached_urls, authenticator_name=None,
                 mdprovider_name=None, key_items=None, value_items=None,
                 secret=None, ttl=None):
        if authenticator_name is None and mdprovider_name is None:
            msg = "You must specify authenticator_name and/or "\
                  "mdprovider_name (otherwise this plugin won't "\
                  "actually do anything...)"
            raise ValueError(msg)
        if secret is None:
            secret = os.urandom(16)
        if ttl is None:
            ttl = 60
        self.memcached_urls = memcached_urls
        self.authenticator_name = authenticator_name
        self.mdprovider_name = mdprovider_name
        self.key_items = key_items
        self.value_items = value_items
        self.secret = secret
        self.ttl = ttl
        self._client = pylibmc.Client(memcached_urls)

    def authenticate(self, environ, identity):
        """Authenticate the identity from the given request environment.

        This method checks whether an entry matching the identity is currently
        stored in the cache.  If so, it loads data from the stored entry and
        return successfully.
        """
        # Only do anything if we're wrapping an authenticator.
        if self.authenticator_name is None:
            return None
        api = get_api(environ)
        if api is None:
            return None
        wrapped_authenticator = api.name_registry[self.authenticator_name]
        # Check if we've got cached data already.
        data = self._get_cached(environ, identity, "authenticate")
        if data is not None:
            identity.update(data)
            return identity.get("repoze.who.userid")
        # Not cached, check with the wrapped authenticator.
        value_items = self.value_items
        if value_items is None:
            old_keys = set(identity.iterkeys())
        userid = wrapped_authenticator.authenticate(environ, identity)
        if userid is None:
            return None
        # If that was successful, cache it along with any added data.
        # Make sure to always cache repoze.who.userid.
        if value_items is None:
            value_items = [k for k in identity.iterkeys() if k not in old_keys]
        identity.setdefault("repoze.who.userid", userid)
        value_items = ["repoze.who.userid"] + list(value_items)
        self._set_cached(environ, identity, value_items, "authenticate")
        return userid

    def add_metadata(self, environ, identity):
        """Add metadata to the given identity dict."""
        # Only do anything if we're wrapping an mdprovider.
        if self.mdprovider_name is None:
            return None
        api = get_api(environ)
        if api is None:
            return None
        wrapped_mdprovider = api.name_registry[self.mdprovider_name]
        # Check if we've got cached data already.
        data = self._get_cached(environ, identity, "add_metadata")
        if data is not None:
            identity.update(data)
            return None
        # Not cached, check with the wrapped mdprovider.
        value_items = self.value_items
        if value_items is None:
            old_keys = set(identity.iterkeys())
        wrapped_mdprovider.add_metadata(environ, identity)
        # Cache any data that was added.
        if value_items is None:
            value_items = [k for k in identity.iterkeys() if k not in old_keys]
        self._set_cached(environ, identity, value_items, "add_metadata")

    def _get_cached(self, environ, identity, method_name):
        """Get the cached data for the given identity, if any.

        If the given identity has previously been seen and stored in the
        cache, this method returns the dict of cached data.  If the identity
        is not cached then it returns None.
        """
        key = self._get_cache_key(environ, identity, method_name)
        # Grab it from the cache as a JSON string.
        try:
            value = self._client.get(key)
        except pylibmc.Error:
            return None
        # Parse it into a dict of data.
        try:
            value = json.loads(value)
        except ValueError:
            return None
        return value

    def _set_cached(self, environ, identity, value_items, method_name):
        """Set the cached data for the given identity.

        This method extracted the named value_items from the identity and
        stores them in the cache for future reference.  It ignores errors
        from writing to the cache.
        """
        key = self._get_cache_key(environ, identity, method_name)
        data = {}
        for name in value_items:
            try:
                data[name] = identity[name]
            except KeyError:
                pass
        try:
            self._client.set(key, json.dumps(data), time=self.ttl)
        except pylibmc.Error:
            pass
        
    def _get_cache_key(self, environ, identity, method_name):
        """Get the key under which to cache the given identity.

        The cache key is a hmac over all the items named in self.key_items
        if specified, or over all the values in sorted key order.  Using a
        hmac ensures that each identity has a unique key while not leaking
        credential information into memcached.

        A single instance of this plugin might cache results for both an
        authenticator and an mdprovider.  To prevent conflicts the name of
        the calling method is also included in the hash.
        """
        # We cache the key when first calculated, so that we always use the
        # value as it was first seen by the plugin.  This prevents cache
        # misses when some other plugin scribbles on the identity.
        envkey = "repoze.who.plugins.memcached.cache-key-" + method_name
        key = environ.get(envkey)
        if key is not None:
            return key
        # If no list of key items is specified, use all keys in sorted order.
        key_items = self.key_items
        if key_items is None:
            key_items = identity.keys()
            key_items.sort()
        # Hash in the method name and the value of all key items.
        hasher = hmac.new(self.secret, method_name, hashlib.sha1)
        hasher.update("\x00")
        for name in key_items:
            hasher.update(str(identity.get(name, "")))
            hasher.update("\x00")
        key = hasher.hexdigest()
        # Cache it for future reference.
        environ[envkey] = key
        return key


def make_plugin(memcached_urls, authenticator_name=None, mdprovider_name=None,
                key_items=None, value_items=None, secret=None, ttl=None):
    """CachedAuthPlugin helper for loading from ini files."""
    memcached_urls = memcached_urls.split()
    if not memcached_urls:
        raise ValueError("You must specify at least one memcached URL")
    if key_items is not None:
        key_items = key_items.split()
    if value_items is not None:
        value_items = value_items.split()
    if ttl is not None:
        ttl = int(ttl)
    plugin = MemcachedPlugin(memcached_urls, authenticator_name,
                             mdprovider_name,key_items, value_items,
                             secret, ttl)
    return plugin
